pageserver/src/tenant/secondary/heatmap_uploader.rs

```diff
@@ -347,12 +344,25 @@ enum UploadHeatmapError {
     Upload(#[from] anyhow::Error),
 }
 
+/// Digests describing the heatmap we most recently uploaded successfully.
+///
+/// md5 is generally a bad hash.  We use it because it's convenient for interop with AWS S3's ETag,
+/// which is also an md5sum.
+#[derive(Clone)]
+struct LastUploadState {
+    // Digest of json-encoded HeatMapTenant
+    uploaded_digest: md5::Digest,
+
+    // Digest without atimes set.
+    layers_only_digest: md5::Digest,
+}
+
 /// The inner upload operation.  This will skip if `last_digest` is Some and matches the digest
 /// of the object we would have uploaded.
 async fn upload_tenant_heatmap(
```

Comment:
It is true that each tenant uploads new heatmap every 60 seconds by default? Isn't it causing too much load?

Reply:
The default is 60s in the storage controller, but currently that only manages a smaller number of large tenants.  For the general population of tiny tenants, a 5 minute period is used (set in control plane).  `heatmap_upload_concurrency` also reduces risk of this interfering with any other work.